DatadogInternal/Sources/NetworkInstrumentation/URLSession/ImmutableRequest.swift

```diff
@@ -0,0 +1,44 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-Present Datadog, Inc.
+ */
+
+import Foundation
+
+/// An immutable version of `URLRequest`.
+///
+/// Introduced in response to concerns raised in https://github.com/DataDog/dd-sdk-ios/issues/1638
+/// it makes a copy of request attributes, safeguarding against potential thread safety issues arising from concurrent
+/// mutations (see more context in https://github.com/DataDog/dd-sdk-ios/pull/1767 ).
+public struct ImmutableRequest {
+    /// The URL of the request.
+    public let url: URL?
+    /// The HTTP method of the request.
+    public let httpMethod: String?
+    /// Known HTTP header fields of the request.
+    public let knownHTTPHeaderFields: [String: String]
+    /// A reference to the original `URLRequest` object provided during initialization. Direct use is discouraged
+    /// due to thread safety concerns. Instead, necessary attributes should be accessed through `ImmutableRequest` fields.
+    public let unsafeOriginal: URLRequest
+
+    public init(request: URLRequest) {
+        self.url = request.url
+        self.httpMethod = request.httpMethod
+
+        // As observed in https://github.com/DataDog/dd-sdk-ios/issues/1638, accessing `request.allHTTPHeaderFields` is not
+        // safe and leads to crashes with undefined root cause. To avoid this, instead we use `request.value(forHTTPHeaderField:)`
+        // to only read headers known by the SDK.
+        var knownHTTPHeaderFields: [String: String] = [:]
+        addHeaderIfExists(request: request, field: TracingHTTPHeaders.originField, to: &knownHTTPHeaderFields)
```

Comment:
💡 We could consider adding other known headers here, notably the ones related to distributed tracing (`x-datadog-*`, W3C and B3). Not doing this as these headers are not required for upstream code. The only reason for `allHTTPHeaderFields` existence, was accessing the `x-datadog-origin` header which is added here.

Alternatively, we could replace the entire `knownHTTPHeaderFields` prop with `datadogOriginField`. This way we will move even further away from the unsafe consideration of "all headers".

Open for discussion 🧠.

Reply:
> Can we copy all headers for instance?

I don't see any option without using the problematic `allHTTPHeaderFields` API.

> This seems risky, by only looking for some headers. We will be blind where knownHTTPHeaderFields is actually used.

Makes sense 👍. I changed it to the alternative approach by replacing `knownHTTPHeaderFields` with `ddOriginHeaderValue` prop.

Reply:
This is those places where I prefer using assertions (debug only) but not sure how customer can selectively disable them for libraries only else it is too much of noise. 

I'm okay.